File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> nums_map;
        stack<int> nums_st;
        vector<int> res(nums1.size(), -1);
        for (int i = 0; i < nums1.size(); ++i) {
            nums_map[nums1[i]] = i;
        }
        
        for (int i = 0; i < nums2.size(); ++i) {
            while (!nums_st.empty() && nums_st.top() < nums2[i]) {
                int curr = nums_st.top();
                nums_st.pop();
                
                if (nums_map.find(curr) != nums_map.end()) {
                    res[nums_map[curr]] = nums2[i]; 
                }
            }
            nums_st.push(nums2[i]);
        }
        return res;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums_dict = {val: idx for idx, val in enumerate(nums1)}
        res = [-1] * len(nums1)

        nums_stack = []
        for i in range(len(nums2)):
            while len(nums_stack) != 0 and nums_stack[-1] < nums2[i]:
                val = nums_stack.pop()
                idx = nums_dict[val]
                res[idx] = nums2[i]
            if nums_dict.get(nums2[i]) is not None:
                nums_stack.append(nums2[i])
        return res
                
                
                
            